distributed/protocol/serialize.py

```diff
@@ -71,29 +74,111 @@ def pickle_loads(header, frames):
     return pickle.loads(x, buffers=buffers)
 
 
+def import_allowed_module(name):
+    if name in _cached_allowed_modules:
+        return _cached_allowed_modules[name]
+
+    # Check for non-ASCII characters
+    name = name.encode("ascii").decode()
+    # We only compare the root module
+    root = name.split(".", 1)[0]
+
+    # Note, if an empty string creeps into allowed-imports it is disallowed explicitly
+    if root and root in dask.config.get("distributed.scheduler.allowed-imports"):
+        _cached_allowed_modules[name] = importlib.import_module(name)
+        return _cached_allowed_modules[name]
+    else:
+        raise RuntimeError(
+            f"Importing {repr(name)} is not allowed, please add it to the list of "
+            "allowed modules the scheduler can import via the "
+            "distributed.scheduler.allowed-imports configuration setting."
+        )
+
+
 def msgpack_decode_default(obj):
     """
-    Custom packer/unpacker for msgpack to support Enums
+    Custom packer/unpacker for msgpack
     """
     if "__Enum__" in obj:
-        mod = importlib.import_module(obj["__module__"])
-        enum_type = getattr(mod, obj["__name__"])
-        obj = getattr(enum_type, obj["name"])
+        mod = import_allowed_module(obj["__module__"])
+        typ = getattr(mod, obj["__name__"])
+        return getattr(typ, obj["name"])
+
+    if "__Set__" in obj:
+        return set(obj["as-list"])
+
+    if "__SubgraphCallable__" in obj:
+        mod = import_allowed_module(obj["__module__"])
+        typ = getattr(mod, obj["__name__"])
+        return typ(*obj["args"])
+
+    if "__Layer__" in obj:
+        obj_name = obj["__name__"]
+        if obj_name == "BasicLayer":
+            # The default implemention of Layer returns a BasicLayer, which might
+            # not be defined in `mod` therefore we import it explicitly here
+            typ = BasicLayer
+        else:
+            mod = import_allowed_module(obj["__module__"])
+            typ = getattr(mod, obj["__name__"])
+        return typ(*obj["args"])
+
+    if "__HighLevelGraph__" in obj:
+        return HighLevelGraph(
+            obj["layers"],
+            obj["dependencies"],
+        )
+
+    if "__Serialized__" in obj:
+        return Serialized(*obj["data"])
+
     return obj
 
 
 def msgpack_encode_default(obj):
     """
-    Custom packer/unpacker for msgpack to support Enums
+    Custom packer/unpacker for msgpack
     """
 
+    if isinstance(obj, Serialize):
+        return {"__Serialized__": True, "data": serialize(obj.data)}
```

Comment:
Good to see this as it should help with some initial test failures over in https://github.com/dask/distributed/pull/4140